scsi/cmd.ixx

```diff
@@ -446,4 +446,55 @@ SPTD::Status cmd_get_configuration(SPTD &sptd)
     return status;
 }
 
+
+export SPTD::Status cmd_kreon_get_security_sector(SPTD &sptd, std::vector<uint8_t> &response_data)
```

Comment:
There are some quirks but I think that **should** be possible, something like `cmd_read_disc_structure(response_data, 0, 0xFF02FDFF, 0xFE, 0, 0)` should cover most of the issues. We just need to know
* Do we need assign allocation length? 4096 > 2048 so I think it should be fine as long as the command doesn't explicitly require 0x800 to be in the command
* Do we need to set control byte? This is from the KREON documentation so it may be a required part of the command sequence
* How to pass `reserved2` data

Reply:
Okay I did some testing with `sg_raw` on linux and it looks like it needs to match the sequence exactly, I tried 
```
AD 00 FF 02 FD FF FE 00 10 00 01 00 - 4096, no ctrl (all zero response)
AD 00 FF 02 FD FF FE 00 10 00 01 C0 - 4096, ctrl byte (all zero response)
AD 00 FF 02 FD FF FE 00 08 00 01 00 - 2048, no ctrl byte (all zero response)
AD 00 FF 02 FD FF FE 00 08 00 01 00 - 2048, ctrl byte (**get response**)
```
so it looks like we would need to pass the size along as well

Reply:
Where would `cmd_kreon_get_security_sector` best be located if it calls `cmd_read_disc_structure`

Reply:
We might be able to do something like this. By adding default values we could avoid breaking existing code. I am not sure if casing on the control byte is the best solution but we could probably make that more robust.
```diff
diff --git a/scsi/cmd.ixx b/scsi/cmd.ixx
index d4c8cac..207b2e8 100644
--- a/scsi/cmd.ixx
+++ b/scsi/cmd.ixx
@@ -69,9 +69,11 @@ SPTD::Status cdb_send_receive(SPTD &sptd, std::vector<uint8_t> &response, T &cdb
 
     // some drives expect CSS related calls to be completed in one go
     // otherwise SCSI error is returned, keep this value high
-    constexpr uint16_t initial_size = 4096;
-
-    response.resize(initial_size);
+    if(cdb.control != 0xC0)
+    {
+        constexpr uint16_t initial_size = 4096;
+        response.resize(initial_size);
+    }
     *(uint16_t *)cdb.allocation_length = endian_swap<uint16_t>(response.size());
     status = sptd.sendCommand(&cdb, sizeof(cdb), response.data(), (uint32_t)response.size());
     if(status.status_code)
@@ -254,7 +256,7 @@ export SPTD::Status cmd_read_cd_text(SPTD &sptd, std::vector<uint8_t> &cd_text)
 
 
 export SPTD::Status cmd_read_disc_structure(SPTD &sptd, std::vector<uint8_t> &response_data, uint8_t media_type, uint32_t address, uint8_t layer_number, READ_DISC_STRUCTURE_Format format,
-    uint8_t agid)
+    uint8_t agid, uint8_t control = 0, uint8_t reserved2 = 0)
 {
     SPTD::Status status;
 
@@ -268,6 +270,9 @@ export SPTD::Status cmd_read_disc_structure(SPTD &sptd, std::vector<uint8_t> &re
     cdb.format = (uint8_t)format;
     cdb.agid = agid;
 
+    cdb.control = control;
+    cdb.reserved2 = reserved2;
+
     status = cdb_send_receive(sptd, response_data, cdb);
 
     return status;
     ```

Reply:
Given the changes required I would suggest leaving `cmd_read_disc_structure` alone and just making `cmd_kreon_get_security_sector` atomic

Reply:
Example of atomic `cmd_kreon_get_security_sector`
```cpp
export SPTD::Status cmd_kreon_get_security_sector(SPTD &sptd, std::vector<uint8_t> &response_data, uint8_t ss_val)
{
    SPTD::Status status;

    // AD 00 FF 02 FD FF FE 00 08 00 xx C0
    CDB12_ReadDiscStructure cdb = {};
    cdb.operation_code = (uint8_t)CDB_OperationCode::READ_DISC_STRUCTURE;
    *(uint32_t *)cdb.address = endian_swap<uint32_t>(0xFF02FDFF);
    cdb.layer_number = 0xFE;
    *(uint16_t *)cdb.allocation_length = endian_swap<uint16_t>((uint16_t)response_data.size());
    cdb.control = 0xC0;
    cdb.reserved2 = ss_val;

    return sptd.sendCommand(&cdb, sizeof(cdb), response_data.data(), response_data.size());
}
```

Reply:
Yeah, I'll loop it in `xbox.ixx` and have it return a bool to show if it is partial or complete. If we want to implement it into `cmd_read_disc_structure` it will be basically the same wrapper function

Reply:
I don't think it is worthwhile to implement it into `cmd_read_disc_structure`, it is different enough.

Reply:
I pulled the loop out and moved it to `xbox.ixx`, now if there is a partial security sector it will attempt to continue the dump but not clean the SS (will also log that it is partial security sector). It still uses the function I added but would be easy enough to adjust it to `cmd_read_disc_structure` if we want to go that route

Reply:
Yes, I agree to keep cmd_kreon_get_security_sector() as it is, the kreon implementation is very hardcoded based on your tests.